docs/architecture/adr-020-protobuf-client-encoding.md

```diff
@@ -0,0 +1,172 @@
+
+Since the messages that an application is known and allowed to handle are specific
+to the application itself, so must the transactions be specific to the application
+itself. Similar to how we described in [ADR 019](./adr-019-protobuf-state-encoding.md),
+the concrete types will be defined at the application level via Protobuf `oneof`.
+
+The application will define a single canonical `Message` Protobuf message
+with a single `oneof` that implements the SDK's `Msg` interface.
+
+Example:
+
+```protobuf
+// app/codec/codec.proto
+
+message Message {
+  option (cosmos_proto.interface_type) = "github.com/cosmos/cosmos-sdk/types.Msg";
+
+  oneof sum {
+    bank.MsgSend = 1;
+    staking.MsgCreateValidator = 2;
+    staking.MsgDelegate = 3;
+    // ...
+  }
+}
+```
+
+Because an application needs to define it's unique `Message` Protobuf message, it
+will by proxy have to define a `Transaction` Protobuf message that encapsulates this
+`Message` type. The `Transaction` message type must implement the SDK's `Tx` interface.
+
+Example:
+
+```protobuf
+// app/codec/codec.proto
+
+message Transaction {
+  option (cosmos_proto.interface_type) = "github.com/cosmos/cosmos-sdk/types.Tx";
+
+  StdTxBase base = 1;
+  repeated Message msgs = 2;
+}
+```
+
+Note, the `Transaction` type includes `StdTxBase` which will be defined by the SDK
+and includes all the core field members that are common across all transaction types.
+Developers do not have to include `StdTxBase` if they wish, so it is meant to be
+used as an auxiliary type.
+
+### Signing
+
+Signing of a `Transaction` must be canonical across clients and binaries. In order
+to provide canonical representation of a `Transaction` to sign over, clients must
+obey the following rules:
+
+- Encode `StdSignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
+  - Default and zero values must be stripped from the output (`0`, `“”`, `null`, `false`, `[]`, and `{}`).
+- Generate canonical JSON to sign via the [JSON Canonical Form Spec](https://gibson042.github.io/canonicaljson-spec/).
+  - This spec should be trivial to interpret and implement in any language.
+
+```Protobuf
+// app/codec/codec.proto
+
+message StdSignDoc {
+  StdSignDocBase base = 1;
+  repeated Message msgs = 2;
+}
+```
+
+### CLI, REST, & Querying
+
+Currently, the queriers, REST, CLI and handlers encode and decode types via Amino
+JSON encoding using a concrete Amino codec. Being that some of the types dealt with
+in the client can be interfaces, similar to how we described in [ADR 019](./adr-019-protobuf-state-encoding.md),
+the client logic will now need to take a codec interface that knows not only how
+to handle all the types, but also knows how to generate transactions, signatures,
+and messages.
+
+```go
+type TxGenerator interface {
+  sdk.Tx
+
+  NewTx() sdk.Tx
+  SetMsgs(...sdk.Msg) error
+  GetSignatures() []StdSignature
+  SetSignatures(...StdSignature) error
+}
```

Comment:
Not sure why `TxGenerator` both has builder functions and a `NewTx` functions. Also we need a way to generate sign bytes.

How about something like this?
```suggestion
type TxGenerator interface {
  NewTx() TxBuilder
  TxSignBytes func(chainID string, accnum uint64, sequence uint64, fee StdFee, msgs []sdk.Msg, memo string) ([]byte, error)
}

type TxBuilder interface {
  std.Tx
  
  SetMsgs(...Msg) error
  SetSignatures(...StdSignature) error
  SetFee(StdFee)
  SetMemo(string)
  
  Marshal() ([]byte, error)
}
```

Reply:
Yeah, good catch. I kind of slapped this together before the weekend, already realizing this is not the correct abstraction. I will adjust 👍